src/libsync/discovery.cpp

```diff
@@ -0,0 +1,1218 @@
+            }
+
+            if (noServerEntry) {
+                item->_instruction = CSYNC_INSTRUCTION_REMOVE;
+                item->_direction = SyncFileItem::Down;
+            }
+        } else if (!dbEntry.isValid() && localEntry.isVirtualFile && !noServerEntry) {
+            // Somehow there is a missing DB entry while the virtual file already exists.
+            // The instruction should already be set correctly.
+            ASSERT(item->_instruction == CSYNC_INSTRUCTION_UPDATE_METADATA);
+            ASSERT(item->_type == ItemTypeVirtualFile)
+            ASSERT(item->_file.endsWith(_discoveryData->_syncOptions._virtualFileSuffix));
+            item->_file.chop(_discoveryData->_syncOptions._virtualFileSuffix.size());
+        } else if (dbEntry.isValid() && !typeChange && ((dbEntry._modtime == localEntry.modtime && dbEntry._fileSize == localEntry.size) || (localEntry.isDirectory && dbEntry._type == ItemTypeDirectory))) {
+            // Local file unchanged.
+            if (noServerEntry) {
+                item->_instruction = CSYNC_INSTRUCTION_REMOVE;
+                item->_direction = SyncFileItem::Down;
+            } else if (!serverModified && dbEntry._inode != localEntry.inode) {
+                item->_instruction = CSYNC_INSTRUCTION_UPDATE_METADATA;
+                item->_direction = SyncFileItem::Down; // Does not matter
+            }
+        } else if (serverModified || dbEntry._type == ItemTypeVirtualFile) {
+            if (serverEntry.isDirectory && localEntry.isDirectory) {
+                // Folders of the same path are always considered equals
+                item->_instruction = CSYNC_INSTRUCTION_UPDATE_METADATA;
+            } else {
+                // It could be a conflict even if size and mtime match!
+                //
+                // In older client versions we always treated these cases as a
+                // non-conflict. This behavior is preserved in case the server
+                // doesn't provide a content checksum.
+                //
+                // When it does have one, however, we do create a job, but the job
+                // will compare hashes and avoid the download if possible.
+                QByteArray remoteChecksumHeader = serverEntry.checksumHeader;
+                if (!remoteChecksumHeader.isEmpty()) {
+                    // Do we have an UploadInfo for this?
+                    // Maybe the Upload was completed, but the connection was broken just before
+                    // we recieved the etag (Issue #5106)
+                    auto up = _discoveryData->_statedb->getUploadInfo(path._original);
+                    if (up._valid && up._contentChecksum == remoteChecksumHeader) {
+                        // Solve the conflict into an upload, or nothing
+                        item->_instruction = up._modtime == localEntry.modtime && up._size == localEntry.size
+                            ? CSYNC_INSTRUCTION_UPDATE_METADATA : CSYNC_INSTRUCTION_SYNC;
```

Comment:
Using an UPDATE_METADATA here is confusing when the below comment says we can't use a typical UPDATE_METADATA and then updates the journal on the spot. Please explain why it isn't a NONE here?

Reply:
I don't know. If the test is very thorough then yes, but probably we don't check the journal state in detail. An extra UPDATE_METADATA shouldn't hurt, it just confused me while reading. Since it was that way before I'd leave it as-is for now, double check later.